lib/Sema/TypeCheckAttr.cpp

```diff
@@ -2290,63 +2290,29 @@ void AttributeChecker::visitDifferentiableAttr(DifferentiableAttr *attr) {
-  SmallVector<TupleTypeElt, 8> retElts;
+  // Validate the 'wrt:' parameters.
 
-  // If `type` is not allowed as a wrt type, diagnoses and returns true.
-  auto checkAndDiagnoseWrtType = [&](SourceLoc loc, Type type) -> bool {
-    if (type->isAnyClassReferenceType() || type->isExistentialType()) {
-      TC.diagnose(
-          loc,
-          diag::differentiable_attr_cannot_diff_wrt_objects_or_existentials,
-          type);
-      return true;
-    }
-    return false;
-  };
-
-  // If the self type of `original` is allowed as a wrt param, appends the
-  // corresponding return type to `retElts` and returns false. Otherwise,
-  // returns true and diagnoses.
-  auto addWrtSelfRetTyOrDiagnose = [&](SourceLoc loc) -> bool {
-    auto parent = original->getParent();
-    if (checkAndDiagnoseWrtType(loc, parent->getSelfTypeInContext()))
-      return true;
-    retElts.push_back(parent->getSelfInterfaceType());
-    return false;
-  };
+  // These are the wrt param indices specified by the user, which have not yet
+  // been checked.
+  auto uncheckedWrtParamIndices = attr->getParameters();
 
-  // If `param` is allowed as a wrt param, appends the corresponding return type
-  // to `retElts` and returns false. Otherwise, returns true and diagnoses.
-  auto addWrtParamRetTyOrDiagnose = [&](SourceLoc loc,
-                                        const ParamDecl *param) -> bool {
-    if (checkAndDiagnoseWrtType(loc, param->getType()))
-      return true;
-    retElts.push_back(param->getInterfaceType());
-    return false;
-  };
+  // We will put the checked wrt param indices here.
+  auto *checkedWrtParamIndices = AutoDiffParameterIndices::create(
+      ctx, originalFnTy,
+      /*isMethod*/original->getImplicitSelfDecl() ? true : false);
 
-  // When 'wrt:' is not specified, the adjoint's return type is the type of all
-  // of original's parameters. The self parameter is intentionally excluded.
-  if (wrtParams.empty()) {
-    auto attrLoc = attr->getLocation();
-    for (auto *param : originalParams)
-      if (addWrtParamRetTyOrDiagnose(attrLoc, param))
-        return;
-  }
-  // If 'wrt:' is specified, make sure it's valid and compute the corresponding
-  // adjoint return type.
-  else {
-    // This helps determine if the parameter indices are ascending.
+  if (uncheckedWrtParamIndices.empty()) {
+    // If 'wrt:' is not specified, the wrt parameters are all the parameters in
+    // the main parameter group. Self is intentionally excluded.
+    checkedWrtParamIndices->addAllNonSelfParams();
+  } else {
+    // 'wrt:' is specified. Validate and collect the selected parameters.
     int lastIndex = -1;
-    // Verify each parameter in 'wrt:' list and collect return types to
-    // `retElts`.
-    for (size_t i = 0; i < wrtParams.size(); i++) {
-      auto paramLoc = wrtParams[i].getLoc();
-      switch (wrtParams[i].getKind()) {
+    for (size_t i = 0; i < uncheckedWrtParamIndices.size(); i++) {
```

Comment:
Seems this can be further simplified to use
```suggestion
    for (auto &param : uncheckedWrtParams) {
```